### DISCUSSION TITLE
DIY Shades with Tasmota on NodeMCU; Problems with upper Endstop/Rules Trigger

### DISCUSSION THREAD

**Discussion by Jannik698** [arendst/Tasmota#22145] [Q&A] [answered]
Hi there. Searched the web for two days now. Not finding any solutions for my problem. Hence this post.

Im building a indoor roller shades for my bedroom. It consists of a geared DC Motor, a NodeMCU with Tasmota plugged in a Motor Driver Shield. On the Motor Driver Shield there is a L293D H-Bridge and some additional circuitry (inverters).

The Module Configuration looks like this:

D3/GPIO0 Relay(2) //Direction Output to Motor Driver
D2/GPIO4 Button(4) //Pushbutton for Testing
D1/GPIO5 Relay(1) //Enable Output to Motor Driver
D6/GPIO12 Relay(3) Just for testing
D7/GPIO13 Relay(4) Just for testing
D5/GPIO14 Switch(3) IR Distance Sensor used to detect if the Shade is fully open

Currently everything is powered with a LabPSU supplying 7V, a DCDC Module stepping that down to 5V for the 3.3V LDO on the NodeMCU. This gives me the option to use higher Voltage Motors later on.

I want the motors to stop rolling the shade down after a given amount of time and to stop rolling the shade up after the IR Sensor reports the shade is fully rolled up. (Rising Edge from IR Sensor --> Switch(3) changes from OFF to ON)

My idea is to write a rule that is triggered by the IR Sensor that disengages the Enable Relay(1) and sets the Shutter Position to "Open"

The Rule that should do this looks like this: "Rule1 ON Switch3#State = 1 DO ShutterStop ENDON ON Switch3#State = 1 DO ShutterSetOpen ENDON"

However i cant get to trigger the rule either by waiting for Switch3#State to change to OFF (=0) or ON (=1). Only configuring the trigger for a Toggle (=2) results in the rule getting executed on each Signal Edge.

When the IR sensor input is configured as a button and the rule is therefore adopted i got this rule:
"Rule1 ON Button3#State = 2 DO ShutterStop ENDON ON Button3#State = 2 DO ShutterSetOpen ENDON"
to work if the IR Sensor is triggered for a short duration. This is not practical as the sensor can be triggered for several hours e.g. over night.

Any Ideas how to configure the rule successfully to trigger only on the rising edge from the IR Sensor?

**Comment by sfromis** [comment, marked as answer]
As default for `Switch`/`Button` inputs is the "normal" wiring of it connecting to ground when operated (falling edge), maybe try treating the input as inverted? For a `Switch` input, that is done by selecting a `SwitchMode` to match what you want.

There is also a `Button_i` config for inverted, but either way, this type of input is designed for quick pushes, not being held for a long time.

**Comment by Jannik698** [comment]
Miraculously inverting the Switch Input worked! Thank You! The rule gets executed when the Shades clears the IR Sensor.

I fail to understand why previously neither the trigger listend for a falling or rising Edge (Switch3#State = 0 or Switch3#State = 1) but inverting the switch now correctly triggers on one edge. What am i missing? Any Ideas?

**Reply by sfromis** [reply]
Comparing with a light switch, the difference is what constitutes the input being active (switch is "on"). Normal wiring means that it is grounded by the physical switch, going low, when turning on. Inverted is that it is going high when turning on. Both work, but better when you match the config to your hardware, to get the transitions "the right way".

What happened previously may be about what `SwitchMode` config you had, as this a very important thing to set when having a `Switch` input.

**Comment by Jannik698** [comment]
I assumed the default Switch Mode was "Follow Mode" but i just now learned, that the default is "Toggle Mode". The behavior in the WebGUI is very similar but the Input never gets turned 'ON' or 'OFF' only toggled.

With Switchmode3 1 (normal Follow Mode) the rule gets triggered on "Rule1 ON Switch3#State = 1" and
with Switchmode3 2 (inverted Follow Mode) the rule gets triggered on "Rule1 ON Switch3#State = 0"

Both configurations work as intended. Thank again sfromis for your help.
